Documentation/proposals/202407-remote-write.md

```diff
@@ -0,0 +1,128 @@
+
+## Why
+
+The Prometheus remote_write configuration is defined in the Prometheus CRD. Currently, the configuration data generation is the responsibility of cluster admins.
+
+## Goals
+
+The main goal is to enable application developers/operators to self-service the remote write, and configure how the client sends metrics to the remote endpoint.
+
+This means exposing a new CRD to configure Prometheus remote_write configuration.
+
+## Non-goals
+
+We do not cover "RemoteRead" CRD here. We can still implement it later if needed.
+
+## How
+
+### RemoteWrite CRD
+
+The RemoteWrite CRD represents one of the Prometheus remote_write configuration scoped to the resource’s namespace.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: RemoteWrite
+metadata:
+  name: example
+  namespace: default
+spec:
+  url: "https://aps-workspaces.us-west-2.amazonaws.com/workspaces/<workspace id>/api/v1/remote_write"
+  sigv4:
+    region: us-west-2
+    accessKey:
+      # Secret object references will be resolved in the namespace of this remoteWrite object.
+      # In this manifest it is the "default" namespace.
+      name: aws_access
+      key: access_key
+    secretKey:
+      name: aws_access
+      key: secret_key
+```
+
+```go
+package v1alpha1
+
+import (
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type RemoteWrite struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec monitoringv1.RemoteWriteSpec `json:"spec"`
+}
+```
+
+### Prometheus CRD
+
+The Prometheus CRD is extended with 2 new fields (remoteWriteSelector and remoteWriteNamespaceSelector) that define which RemoteWrite resources are associated with this Prometheus instance.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+metadata:
+  name: example
+  namespace: default
+spec:
+  # RemoteWrite objects to be selected. An empty label selector matches all
+  # objects. A null label selector matches no objects.
+  remoteWriteSelector: {}
+  # Namespaces to match for RemoteWrite discovery. An empty label selector
+  # matches all namespaces. A null label selector matches the current namespace
+  # only.
+  remoteWriteNamespaceSelector: {}
+  ...
+```
+
+#### Respect enforceNamespaceLabel and excludedFromEnforcement
+
+The RemoteWrite CRD should respect the Prometheus `.spec.enforcedNamespaceLabel` field.
+
+The object’s namespace is added as the first item in the `write_relabel_config` of the remote_write configuration.
+
+```yaml
+write_relabel_configs:
+- sourceLabels:
+  - namespace
+  regex: <namespace>
+  action: keep
+  ...
+```
+
+The item is inserted in the first position by the operator before any other user relabelings then there's no way that users can have access to "foreign" metrics.
```

Comment:
Pretty neat. But I am still not sure about how reliable this would be.

Reply:
Might be a nit, but for clarity, could we elaborate on what we mean by "foreign" metrics?

Reply:
```suggestion
The item is inserted in the first position by the operator before any other user relabelings, so there's no way that users can have access to metrics labeled with namespaces other than the one in which the RemoteWrite object is created.
```